compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor.hs

```diff
@@ -342,3 +346,130 @@ checkKinds (GHC.L _ m) = do
         [ "Reference to GHC.Types.Symbol will not be preserved during DAML compilation."
         , "This will cause problems when importing this module via data-dependencies."
         ]
+
+-- | Issue #6788: Rewrite multi-binding let expressions,
+--
+--      let x1 = e1
+--          x2 = e2
+--          ...
+--          xn = en
+--      in b
+--
+-- Into a chain of single-binding let expressions,
+--
+--      let x1 = e1 in
+--        let x2 = e2 in
+--          ...
+--            let xn = en in
+--              b
+--
+-- Likewise, rewrite multi-binding let statements in a "do" block,
+--
+--      do
+--          ...
+--          let x1 = e1
+--              x2 = e2
+--              ...
+--              xn = en
+--          ...
+--
+-- Into single-binding let statements,
+--
+--      do
+--          ...
+--          let x1 = e1
+--          let x2 = e2
+--          ...
+--          let xn = en
+--          ...
+--
+rewriteLets :: GHC.ParsedSource -> GHC.ParsedSource
+rewriteLets = fmap onModule
+  where
+    onModule :: GHC.HsModule GHC.GhcPs -> GHC.HsModule GHC.GhcPs
+    onModule x = x { GHC.hsmodDecls = map onDecl $ GHC.hsmodDecls x }
+
+    onDecl :: GHC.LHsDecl GHC.GhcPs -> GHC.LHsDecl GHC.GhcPs
+    onDecl = fmap (Uniplate.descendBi onExpr)
+
+    onExpr :: GHC.LHsExpr GHC.GhcPs -> GHC.LHsExpr GHC.GhcPs
+    onExpr = Uniplate.transform $ \case
+        GHC.L loc (GHC.HsLet GHC.NoExt letBinds letBody) ->
+            foldr (\bind body -> GHC.L loc (GHC.HsLet GHC.NoExt bind body))
+                letBody
+                (orderLocalBinds letBinds)
+
+        GHC.L loc1 (GHC.HsDo GHC.NoExt doContext (GHC.L loc2 doStmts)) ->
+            GHC.L loc1 (GHC.HsDo GHC.NoExt doContext
+                (GHC.L loc2 (concatMap onStmt doStmts)))
+
+        x -> x
+
+    onStmt :: GHC.ExprLStmt GHC.GhcPs -> [GHC.ExprLStmt GHC.GhcPs]
+    onStmt = \case
+        GHC.L loc (GHC.LetStmt GHC.NoExt letBinds) ->
+            map (GHC.L loc . GHC.LetStmt GHC.NoExt) (orderLocalBinds letBinds)
+        x -> [x]
+
+    orderLocalBinds :: GHC.LHsLocalBinds GHC.GhcPs -> [GHC.LHsLocalBinds GHC.GhcPs]
+    orderLocalBinds = \case
+        GHC.L loc (GHC.HsValBinds GHC.NoExt valBinds)
+            | GHC.ValBinds GHC.NoExt bindBag sigs <- valBinds
+            , binds <- GHC.bagToList bindBag
+            , length binds > 1
+            -> makeLocalValBinds loc binds sigs
+        binds -> [binds]
+
+    makeLocalValBinds :: GHC.SrcSpan
+        -> [GHC.LHsBindLR GHC.GhcPs GHC.GhcPs]
+        -> [GHC.LSig GHC.GhcPs]
+        -> [GHC.LHsLocalBinds GHC.GhcPs]
+    makeLocalValBinds _ [] _ = []
```

Comment:
Isn't this dropping signatures without bindings on the floor? "Unbound" signatures happen to me quite regularly because I have a typo in a name somewhere. I would suggest putting all remaining signatures in a final let binding such that we get GHC's errors for them.

Reply:
Good point. I'll add a test as well.

Reply:
I'm confused now. What happens in the `makeLocalValBinds _ [] (sig1:sigs)` case to `sig1`? The test indicates that `sig1` gets preserved but I can't find the piece of code responsible for preserving it.

Reply:
There's no `makeLocalValBinds _ [] _` case ... actually it could be removed since it never gets invoked.

Reply:
The case you're looking for is right below on line 428 (-:

Reply:
I opened a PR to make the non-emptiness of `binds` explicit: https://github.com/digital-asset/daml/pull/8013

Reply:
Strange, somehow GitHub did not show me the third PR, which added line 428. Now it does and it all makes sense. The use of non-empty lists makes it clearer nevertheless. Thank you.